From dc3996d9c777357fd22e20b15ee165ce416a6097 Mon Sep 17 00:00:00 2001
From: Pradeep Kumar Chitrapu <quic_pradeepc@quicinc.com>
Date: Fri, 10 Dec 2021 10:53:05 -0800
Subject: [PATCH 24/27] mac80211: copy MU-MIMO and OFDMA EHT capability flags
 to driver

Pass MU-MIMO and OFDMA configurations set in EHT IEs from user
space to driver.

Signed-off-by: Pradeep Kumar Chitrapu <quic_pradeepc@quicinc.com>
---
 include/net/mac80211.h | 10 ++++++++++
 net/mac80211/cfg.c     | 12 ++++++++++++
 2 files changed, 22 insertions(+)

--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -655,6 +655,12 @@ struct ieee80211_fils_discovery {
  * @he_full_ul_mumimo: does this BSS support the reception (AP) or transmission
  *     (non-AP STA) of an HE TB PPDU on an RU that spans the entire PPDU
  *     bandwidth
+ * @eht_su_beamformer: does this BSS support operation as an EHT SU beamformer
+ * @eht_su_beamformee: does this BSS support operation as an EHT SU beamformee
+ * @eht_mu_beamformer: does this BSS support operation as an EHT MU beamformer
+ * @eht_80mhz_full_ul_mumimo: does this BSS support the reception (AP) or transmission
+ *	(non-AP STA) of an EHT TB PPDU on an RU that spans the entire PPDU
+ *	bandwidth
  * @he_obss_pd: OBSS Packet Detection parameters.
  * @he_bss_color: BSS coloring settings, if BSS supports HE
  * @fils_discovery: FILS discovery configuration
@@ -760,6 +766,10 @@ struct ieee80211_bss_conf {
 		u8 ccfs;
 		u8 present_bm;
 	} eht_oper;
+	bool eht_su_beamformer;
+	bool eht_su_beamformee;
+	bool eht_mu_beamformer;
+	bool eht_80mhz_full_bw_ul_mumimo;
 
 	bool csa_active;
 	bool mu_mimo_owner;
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -1249,6 +1249,18 @@ static int ieee80211_start_ap(struct wip
 			return -EOPNOTSUPP;
 
 		sdata->vif.bss_conf.eht_support = true;
+		sdata->vif.bss_conf.eht_su_beamformer =
+				params->eht_cap->fixed.phy_cap_info[0] &
+				IEEE80211_EHT_PHY_CAP0_SU_BEAMFORMER;
+		sdata->vif.bss_conf.eht_su_beamformee =
+				params->eht_cap->fixed.phy_cap_info[0] &
+				IEEE80211_EHT_PHY_CAP0_SU_BEAMFORMEE;
+		sdata->vif.bss_conf.eht_mu_beamformer =
+				params->eht_cap->fixed.phy_cap_info[7] &
+				IEEE80211_EHT_PHY_CAP7_MU_BEAMFORMER_80MHZ;
+		sdata->vif.bss_conf.eht_80mhz_full_bw_ul_mumimo =
+				params->eht_cap->fixed.phy_cap_info[7] &
+				IEEE80211_EHT_PHY_CAP7_NON_OFDMA_UL_MU_MIMO_80MHZ;
 		ieee80211_eht_op_ie_to_bss_conf(&sdata->vif, params->eht_oper);
 	}
 
